bldenv/man/bldlock.man
#
# @(#)38        1.2  src/bldenv/man/bldlock.man, bldprocess, bos412, GOLDA411a 5/12/92 13:28:09
#
# COMPONENT_NAME: (BLDTOOLS) BAI Build Tools
#
# FUNCTIONS: bldlock man page
#
# ORIGINS: 27
#
# (C) COPYRIGHT International Business Machines Corp. 1991,1992
# All Rights Reserved
# Licensed Materials - Property of IBM
#
# US Government Users Restricted Rights - Use, duplication or
# disclosure restricted by GSA ADP Schedule Contract with IBM Corp.
#

bldlock


PURPOSE

	The bldlock command is a general purpose locking function which can be
used to prevent multiple processes from accessing the same resource 
simultaneously.


SYNOPSIS

	bldlock [-l <string> | -u <string>] [-p <permissions>]
		[-f <lockfile>] [ -t <time>]


DESCRIPTION

	The bldlock command provides sychronization for multiple processes 
trying to share the same resource. It uses a common flat file to achieve this. 
Any process which wants to lock a resource passes a string (e.g - filename 'x')
to the bldlock command which in turn appends an entry of this string to the 
'lockfile'. Any other process trying to access the same resource 'x' has to do 
the same thing, since the first process has already used bldlock to lock (i.e 
added an entry to the 'lockfile') the second process has to wait. In the
meantime, the first process which now has sole authority over the resource does
whatever processing it has to with the resource and then calls bldlock to 
unlock the resource by passing the same string ('x' in this case), which in 
turn removes the entry('x') from the 'lockfile'. The second process which has
been waiting can now lock the resource and continue processing. 

	There are a few points worth noting here: (1) The actual resource is 
locked and unlocked in an indirect way by the entry in the lockfile (2) This 
means any process can open ,write,read etc the actual resource ('x') if the
bldlock command is not used (3) Assuming all the processes co-operate by 
calling the bldlock command, they must unlock when done processing otherwise it
causes a deadlock. The wait period, by default for a process is 10 seconds.This
can be changed by using the -t option. The maximum time a process waits for the
lock is 10 seconds by default before bldlock prints a error message of 'timed 
out' and returns to the calling process with a return code of 2, During this 
period bldlock wakes up every 2 seconds and checks for the lock ( i.e if the 
entry is no longer present in the lockfile). Co-operating processes can specify
their own lockfile by using the -f option, otherwise bldlock tries to pick up 
the lockfile from the environment variable 'LOCKFILE' or defaults to '$BLDTMP
/lockfile'.

	For additional security the -p flag for permissions was added. With 
this option, any process trying to lock can also pass its own permissions 
string along with its locking string. The bldlock command appends this new 
entry containing both the locking and permissions strings, When another 
process tries to unlock, it has to specify locking string and the permissions 
string. 


FLAGS

	-l <string>		Appends an entry <string> to the lockfile

	-u <string>		Removes the entry <string> from the lockfile

	-p <permissions>	Permissions string <permissions>

	-t <time>		The maximum wait period in seconds before 
				timing out

	-f <lockfile>		The file <lockfile> to be used as lockfile


EXAMPLES

	1. To lock v3bld with the permission string and a time out of 300 secs
	
		bldlock -l v3bld -p "$(hostname):$$" -t 300

	2. To unlock v3bld

		bldlock -u v3bld -p "$(hostname):$$"


RELATED INFORMATION

	The following file
		/usr/bin/perl


bldenv/bldtools/bldlock.pl
#! /usr/bin/perl
# @(#)31	1.7  src/bldenv/bldtools/bldlock.pl, bldprocess, bos412, GOLDA411a 5/13/92 08:36:35
#
# COMPONENT_NAME: (BLDPROCESS) BAI Build Process
#
# FUNCTIONS: bldlock
#	     getlock
#	     getindex
#	     read_to_ary
#	     write_to_lock
#	     write_ary_tolock
#	     lock
#	     unlock
#	     process_opt_arg
#	     usage
#	     check_opt
#	     get_cmd_line
#	     get_lockfile
#
# ORIGINS: 27
#
# IBM CONFIDENTIAL -- (IBM Confidential Restricted when
# combined with the aggregated modules for this product)
#                  SOURCE MATERIALS
# (C) COPYRIGHT International Business Machines Corp. 1991
# All Rights Reserved
#
# US Government Users Restricted Rights - Use, duplication or
# disclosure restricted by GSA ADP Schedule Contract with IBM Corp.
#
# NAME: bldlock
#
# FUNCTION: The bldlock command is a general purpose locking function which can
# 	    be used to prevent multiple processes from accessing the same 
#	    resource simultaneously.
#
# INPUT: command (locking and unlocking string), PERM, time, lockfile
#
# OUTPUT: Standard output
#
# EXECUTION ENVIRONMENT: Build process environment
#

$LOCK_EX=2;
$LOCK_UN=8;
$sleep_time = 2;

&get_cmd_line;
$time = 10
	unless defined $time;
&get_lockfile;
if ($locked){
	&lock;
}
else{
	&unlock if ($unlocked);
	exit 0;
}

###############################################################################

sub getlock{
	flock(LOCKFILE,$LOCK_EX);
}

###############################################################################


sub getindex{
	local($cmd) = @_;
	$index = '';
	for (0..$#lock_ary){
		local($lck_str,$perm) = split(/\|/,$lock_ary[$_]);
		if ($lck_str eq $cmd){
			next if (($unlocked) && ($perm ne $PERM));
			$index = $_;
			last;
		}
	}
}

###############################################################################


sub lock{
	open(LOCKFILE,"+>>$lockfile");
	chmod 0666, $lockfile;
	&getlock;
	&read_to_ary;
	&getindex($command);
	seek(LOCKFILE,0,2);
	if ($index =~ /^$/){
		&write_to_lock("$command|$PERM\n");
		exit 0;
	}
	else{
		exit 2 if($time <= 1);
		close(LOCKFILE);
		sleep $sleep_time;
		$time -= $sleep_time;
		&lock;
	}
}

###############################################################################


sub read_to_ary{
	local($buf);
	local($len);
	local($BUFFER) = '';
	while ($len = sysread(LOCKFILE,$buf,16384)){
		$BUFFER .= $buf;
		if (!defined $len){
			next if $! =~ /^Interrupted/;
			die "$0: System read error: $!\n";
		}
	}
	@lock_ary = split(/\n/,$BUFFER);
}

###############################################################################


sub write_to_lock{
	local($buf) = @_;
	$len = length($buf);
	$offset = 0;
	while($len){ 
	        $written = syswrite(LOCKFILE,$buf,$len,$offset);
	        die "$0: System write error: $!\n"
	                unless defined $written;
	        $len -= $written;
	        $offset += $written;
	}
}

###############################################################################

sub write_ary_tolock{
	truncate(LOCKFILE,0);
	seek(LOCKFILE,0,0);
	for (0..$#lock_ary){
		&write_to_lock("$lock_ary[$_]\n");
	}
}

###############################################################################

sub unlock{
	open (LOCKFILE,"+>>$lockfile") ||
		die "$0: Can't find the lock file ";
	die "$0: The lock file is empty" if (-z "$lockfile");
	&getlock;
	&read_to_ary;
	&getindex($command);
	exit 0 if($index =~ /^$/);
	splice(@lock_ary,$index,1);
	&write_ary_tolock;
	flock(LOCKFILE,$LOCK_UN);
}

###############################################################################


# NAME: get_cmd_line
#
# FUNCTION: gets the command line options and their arguments
#
# INPUT: avalue,bvalue,cvalue,dvalue,evalue,filename
#
# OUTPUT: avalue,bvalue,cvalue,dvalue,evalue,filename
#

###############################################################################


sub get_cmd_line{
	while ($ARGV[0] =~ /^-/){
		$_ = shift(@ARGV);
		if (/^-(l)(.*)/){
			&checkopt;
			&process_opt_arg($command);
			$locked = 1;
		}
		elsif (/^-(u)(.*)/){
			&checkopt;
			&process_opt_arg($command);
			$unlocked = 1;
		}
		elsif (/^-(p)(.*)/){
			&process_opt_arg($PERM);
		}
		elsif (/^-(t)(.*)/){
			&process_opt_arg($time);
		}
		elsif (/^-(f)(.*)/){
			&process_opt_arg($lockfile);
		}
		else {
			print "Unrecognized option: $_\n";
			&usage;
		}
	}
}

###############################################################################


#
# NAME: process_opt_arg
#
# FUNCTION: Processes the command line optional argument
#
# INPUT: ARGV (global special perl variable)
#	 $1 $2 (Special perl local variables)
#	 $_ (special local perl variable)
#
# OUTPUT: ARGV (global special perl variable)
#	  $_ (special local perl variable)
#

sub process_opt_arg{
	local($option) = $1;
	local($temp);
	$_[0] = ( $2 ? $2 : shift(@ARGV));
	die "$0: $option requires a value\n" if $_[0] =~ /^-|^$/;
}

###############################################################################


sub checkopt{
	die "$0: locking and unlocking are mutually exclusive" 
		if (($locked) || ($unlocked));
}


###############################################################################

sub usage{
	local($line);
	$line = "$0 [-l <lock string> | -u <unlock string>] [-p <string>]";
	$line .= " [-t <seconds>]\n	  [-f <lockfile>]";
	die "$0: $line\n";
}

###############################################################################


sub get_lockfile{

	if ($lockfile eq ""){
		if ( $ENV{'LOCKFILE'} ne ""){
			$lockfile = $ENV{'LOCKFILE'};
		}
		else{
			$lockfile = "/tmp/lockfile";
		}	
	}
}
